src/Types/TypedTxOut.purs

```diff
@@ -0,0 +1,289 @@
+typedTxOutDatumHash (TypedTxOut { txOut }) = (unwrap txOut).data_hash
+
+-- | Extract the `Value` of a `TypedTxOut`
+typedTxOutValue
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => TypedTxOut a b
+  -> Value
+typedTxOutValue (TypedTxOut { txOut }) = (unwrap txOut).amount
+
+-- | Extract the `TxOut` ~ `TransactionOutput` of a `TypedTxOut`
+typedTxOutTxOut
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => TypedTxOut a b
+  -> TransactionOutput
+typedTxOutTxOut (TypedTxOut { txOut }) = txOut
+
+-- Purescript's lack of dependent types requires this smart constructor as we
+-- cannot constrain the `TypedTxOut` datatype.
+-- | Smart constructor to create a `TypedTxOut` from a network ID,
+-- | a correctly-typed data, script, an address, and a value. A smart
+-- | constructor is required because extra constraints are needed.
+-- | `TransactionOutput` is tagged by a phantom type.
+mkTypedTxOut
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => NetworkId
+  -> TypedValidator a
+  -> b
+  -> Value
+  -> Maybe (TypedTxOut a b)
+mkTypedTxOut networkId typedVal dt amount = do
+  dHash <- datumHash $ Datum $ toData dt
+  let address = typedValidatorAddress networkId typedVal
+  pure $ mkTypedTxOut' (wrap { address, amount, data_hash: pure dHash }) dt
+  where
+  mkTypedTxOut'
+    :: TransactionOutput
+    -> b -- Data
+    -> TypedTxOut a b
+  mkTypedTxOut' txOut dat = TypedTxOut { txOut, data: dat }
+
+data WrongOutTypeError = ExpectedScriptGotPubkey -- | ExpectedPubkeyGotScript
+
+derive instance Generic WrongOutTypeError _
+derive instance Eq WrongOutTypeError
+derive instance Ord WrongOutTypeError
+
+instance Show WrongOutTypeError where
+  show = genericShow
+
+-- | An error we can get while trying to type an existing transaction part.
+data ConnectionError
+  = WrongValidatorAddress Address Address
+  | WrongOutType WrongOutTypeError
+  | WrongRedeemerType PlutusData
+  | WrongDatumType PlutusData
+  | CannotConvertOgmiosAddress OgmiosAddress
+  | CannotConvertOgmiosDatumHash String
+  | CannotQueryDatum DatumHash
+  | CannotMakeTypedTxOut
+  | UnknownRef
+
+derive instance Generic ConnectionError _
+derive instance Eq ConnectionError
+derive instance Ord ConnectionError
+
+instance Show ConnectionError where
+  show = genericShow
+
+-- | Checks that the given validator hash is consistent with the actual validator.
+checkValidatorAddress
+  :: forall (a :: Type) (m :: Type -> Type)
+   . Monad m
+  => NetworkId
+  -> TypedValidator a
+  -> Address
+  -> m (Either ConnectionError Unit)
+checkValidatorAddress networkId typedVal actualAddr = runExceptT do
+  let expectedAddr = typedValidatorAddress networkId typedVal
+  unless (expectedAddr == actualAddr)
+    $ throwError
+    $ WrongValidatorAddress expectedAddr actualAddr
+
+-- -- | Checks that the given redeemer script has the right type.
+-- checkRedeemer
```

Comment:
Do we need this? Is this related to the note on `addOwnInput`?

Reply:
Yes, it's related to the fact they use `TxIn` in `addOwnInput` which adds a redeemer which we don't. I can remove if you want.